include/boost/geometry/algorithms/detail/buffer/get_piece_turns.hpp

```diff
@@ -34,34 +34,80 @@ namespace boost { namespace geometry
 namespace detail { namespace buffer
 {
 
-
-#if defined(BOOST_GEOMETRY_BUFFER_USE_SIDE_OF_INTERSECTION)
-struct buffer_assign_turn
+template <typename Ring>
+struct unique_sub_range_from_piece
 {
-    static bool const include_no_turn = false;
-    static bool const include_degenerate = false;
-    static bool const include_opposite = false;
-
-    template
-    <
-        typename Info,
-        typename Point1,
-        typename Point2,
-        typename IntersectionInfo
-    >
-    static inline void apply(Info& info,
-                             Point1 const& /*p1*/,
-                             Point2 const& /*p2*/,
-                             IntersectionInfo const& iinfo)
+    typedef typename boost::range_iterator<Ring const>::type iterator_type;
+    typedef typename geometry::point_type<Ring const>::type point_type;
+
+    unique_sub_range_from_piece(Ring const& ring, iterator_type it0, iterator_type it1)
+        : m_ring(ring)
+        , m_point_i(*it0)
+        , m_point_j(*it1)
+        , m_iterator(it1)
+        , m_point_retrieved(false)
+    {}
+
+    static inline bool is_first() { return false; }
+
+    static inline std::size_t size() { return 3u; }
+
+    inline point_type const& at(std::size_t index) const
+    {
+        switch (index)
+        {
+            case 0 : return m_point_i;
+            case 1 : return m_point_j;
+            case 2 : return get_next_point();
+            default : return m_point_i;
+        }
+    }
+
+private :
+
+    inline point_type const& get_next_point() const
     {
-        info.rob_pi = iinfo.rpi();
-        info.rob_pj = iinfo.rpj();
-        info.rob_qi = iinfo.rqi();
-        info.rob_qj = iinfo.rqj();
+        if (! m_point_retrieved)
+        {
+            m_point = *next_point(m_iterator);
+            m_point_retrieved = true;
+        }
+        return m_point;
+    }
+
+private :
+
+    inline void circular_move_along_ring(iterator_type& next) const
+    {
+        ++next;
+        if (next == boost::end(m_ring))
+        {
+            next = boost::begin(m_ring) + 1;
+        }
     }
 
+    inline iterator_type next_point(iterator_type it) const
+    {
+        iterator_type result = it;
+        circular_move_along_ring(result);
+
+        // TODO: we could also use piece-boundaries
+        // to check if the point equals the last one
+        while (geometry::equals(*it, *result))
+        {
+            circular_move_along_ring(result);
+        }
+        return result;
+    }
+
+    Ring const& m_ring;
+    point_type m_point_i;
+    point_type m_point_j;
+    iterator_type m_iterator;
+
+    mutable point_type m_point;
```

Comment:
Right, the same here. Is there a reason why simply the iterator isn't set to point to the relevant point in a range and left there instead of copying the point here?